### ISSUE TITLE
Missing unnecessary unsafe block warning

### ISSUE THREAD

**Issue by bluss** [rust-lang/rust#45107]
The following code uses a redundant `unsafe` block inside a closure.

This is a warning on stable but there is no warning on nightly.

To reproduce [(playground)](https://play.rust-lang.org/?gist=5533505eeb63832444b27229570838a7&version=nightly):

```rust
fn main() {
    let mut v = Vec::<i32>::with_capacity(24);
    
    unsafe {
        let f = |v: &mut Vec<_>| {
            unsafe {
                v.set_len(24);
            }
        };
        
        v.set_len(0);
        f(&mut v);
    }
}
```

Probably due to #44700, cc @arielb1

**Comment by nikomatsakis** [comment]
triage: P-high

**Comment by arielb1** [comment]
Fixed in #45785.
